chainerrl/agents/ppo.py

```diff
@@ -0,0 +1,265 @@
+            state s  |->  (pi(s, _), v(s))
+        optimizer (chainer.Optimizer): optimizer used to train the model
+        gpu (int): GPU device id if not None nor negative
+        gamma (float): Discount factor [0, 1]
+        lambd (float): Lambda-return factor [0, 1]
+        phi (callable): Feature extractor function
+        value_func_coeff (float): Weight coefficient for loss of
+            value function (0, inf)
+        entropy_coeff (float): Weight coefficient for entropoy bonus [0, inf)
+        update_interval (int): Model update interval in step
+        minibatch_size (int): Minibatch size
+        epochs (int): Training epochs in an update
+        clip_eps (float): Epsilon for pessimistic clipping of likelihood ratio
+            to update policy
+        clip_eps_vf (float): Epsilon for pessimistic clipping of value
+            to update value function. If it is ``None``, value function is not
+            clipped on updates.
+        average_v_decay (float): Decay rate of average V, only used for
+            recording statistics
+        average_loss_decay (float): Decay rate of average loss, only used for
+            recording statistics
+    """
+
+    saved_attributes = ['model', 'optimizer']
+
+    def __init__(self, model, optimizer,
+                 gpu=None,
+                 gamma=0.99,
+                 lambd=0.95,
+                 phi=lambda x: x,
+                 value_func_coeff=1.0,
+                 entropy_coeff=0.01,
+                 update_interval=2048,
+                 minibatch_size=64,
+                 epochs=10,
+                 clip_eps=0.2,
+                 clip_eps_vf=0.2,
+                 average_v_decay=0.999, average_loss_decay=0.99,
+                 ):
+        self.model = model
+
+        if gpu is not None and gpu >= 0:
+            cuda.get_device_from_id(gpu).use()
+            self.model.to_gpu(device=gpu)
+
+        self.optimizer = optimizer
+        self.gamma = gamma
+        self.lambd = lambd
+        self.phi = phi
+        self.value_func_coeff = value_func_coeff
+        self.entropy_coeff = entropy_coeff
+        self.update_interval = update_interval
+        self.minibatch_size = minibatch_size
+        self.epochs = epochs
+        self.clip_eps = clip_eps
+        self.clip_eps_vf = clip_eps_vf
+
+        self.average_v = 0
+        self.average_v_decay = average_v_decay
+        self.average_loss_policy = 0
+        self.average_loss_value_func = 0
+        self.average_loss_entropy = 0
+        self.average_loss_decay = average_loss_decay
+
+        self.xp = self.model.xp
+        self.target_model = None
+        self.last_state = None
+
+        self.memory = []
+        self.last_episode = []
+
+    def _act(self, state, train):
+        xp = self.xp
+        with chainer.using_config('train', train):
+            b_state = batch_states([state], xp, self.phi)
+            action_distrib, v = self.model(b_state)
```

Comment:
Since you won't backprop `action_distrib` nor `v`, it's better to use `chainer.no_backprop_mode` here.